lib/cretonne/src/licm.rs

```diff
@@ -15,78 +22,189 @@ pub fn do_licm(func: &mut Function,
+                        loop_analysis: &LoopAnalysis)
+                        -> DefPosition {
+    let mut closest_def_position = DefPosition::OutsideLoop();
+    for arg in func.dfg.inst_args(inst).into_iter() {
+        let new_def_position = match func.dfg.value_def(*arg) {
+            ValueDef::Arg(ebb_def, _) => {
+                match loop_analysis.base_loop_ebb(ebb_def) {
+                    None => DefPosition::OutsideLoop(),
+                    Some(loop_def) => {
+                        if inst_loop != loop_def {
+                            DefPosition::ParentLoop(loop_def)
+                        } else {
+                            DefPosition::InsideCurrentLoop()
+                        }
+                    }
                 }
-                // If there is a natural pre-header we insert new instructions just before the
-                // related jumping instruction (which is not necessarily at the end).
-                Some((_, last_inst)) => {
-                    pos = Cursor::new(&mut func.layout);
-                    pos.goto_inst(last_inst);
+            }
+            ValueDef::Res(inst_def, _) => {
+                match loop_analysis.base_loop_inst(inst_def, &func.layout) {
+                    None => DefPosition::OutsideLoop(),
+                    Some(loop_def) => {
+                        if inst_loop != loop_def {
+                            DefPosition::ParentLoop(loop_def)
+                        } else {
+                            DefPosition::InsideCurrentLoop()
+                        }
+                    }
+                }
+            }
+        };
+        // We now find some sort of minimum between `closest_def_position` and `new_def_position`.
+        match (closest_def_position.clone(), new_def_position) {
+            (DefPosition::InsideCurrentLoop(), _) |
+            (_, DefPosition::InsideCurrentLoop()) => {
+                closest_def_position = DefPosition::InsideCurrentLoop();
+                break;
+            }
+            (DefPosition::ParentLoop(_), DefPosition::OutsideLoop()) => break,
+            (DefPosition::OutsideLoop(), DefPosition::ParentLoop(new_lp)) => {
+                closest_def_position = DefPosition::ParentLoop(new_lp);
+            }
+            (DefPosition::ParentLoop(old_lp), DefPosition::ParentLoop(new_lp)) => {
+                if loop_analysis.is_child_loop(new_lp, old_lp) {
+                    closest_def_position = DefPosition::ParentLoop(new_lp)
+                } else {
+                    closest_def_position = DefPosition::ParentLoop(old_lp)
                 }
-            };
-            // The last instruction of the pre-header is the termination instruction (usually
-            // a jump) so we need to insert just before this.
-            for inst in invariant_inst {
-                pos.insert_inst(inst);
             }
+            (DefPosition::OutsideLoop(), DefPosition::OutsideLoop()) => (),
+
         }
```

Comment:
In C++ I might expect the code above to look like this:

```
if (new_def_position.kind() < closest_def_position.kind()) {
    closest_def_position = new_def_position;
}
```

instead of a big match/switch on the pair of kind() values. Is something like this possible in Rust?